### PR TITLE
Handshake refactoring.

### PR DESCRIPTION
**Pull request by hdevalence**
This has the refactoring mentioned in #78, specifically https://github.com/ZcashFoundation/zebra2-maybe/pull/78#discussion_r337674531

### CODE REVIEW

**Comment by hdevalence** [comment]
Github renders the `connector.rs` changes as an end-to-end rewrite, but the commits are structured as a move and then a refactor.

File: zebra-network/src/peer/connector.rs [hunk 1]
```diff
@@ -72,173 +46,14 @@ where
...
+    fn call(&mut self, addr: SocketAddr) -> Self::Future {
+        let mut hs = self.handshaker.clone();
+        async move {
+            let stream = TcpStream::connect(addr).await?;
+            hs.ready().await?;
+            let client = hs.call((stream, addr)).await?;
+            Ok(Change::Insert(addr, client))
+        }
+            .boxed()
     }
```
Comment (dconnolly) [inline zebra-network/src/peer/connector.rs:58]: 👍

File: zebra-network/src/peer/server.rs [hunk 2]
```diff
@@ -261,7 +261,10 @@ where
                     .expect("response oneshot should be unused");
                 AwaitingRequest
             }
-            (AwaitingResponse(Ping(req_nonce), tx), Message::Pong(res_nonce)) => {
+            // In this special case, we ignore tx, because we handle ping/pong
+            // messages internally; the "shadow client" only serves to generate
+            // outbound pings for us to process.
+            (AwaitingResponse(Ping(req_nonce), _tx), Message::Pong(res_nonce)) => {
```
Comment (dconnolly) [inline zebra-network/src/peer/server.rs:267]: 👍

File: zebra-network/src/peer_set.rs [hunk 3]
```diff
@@ -75,21 +61,28 @@ where
...
-        ),
-        1,
-    );
+
+    // Construct services that handle inbound handshakes and perform outbound
+    // handshakes. These use the same handshake service internally to detect
+    // self-connection attempts. Both are decorated with a tower TimeoutLayer to
+    // enforce timeouts as specified in the Config.
+    let (listener, connector) = {
+        use tower::timeout::TimeoutLayer;
```
Comment (dconnolly) [inline zebra-network/src/peer_set.rs:70]: ❤️

File: zebra-network/src/peer_set.rs [hunk 4]
```diff
@@ -158,41 +146,33 @@ where
     (peer_set, address_book)
 }
 
-/// Use the provided `peer_connector` to connect to `initial_peers`, then send
+/// Use the provided `handshaker` to connect to `initial_peers`, then send
```
Comment (dconnolly) [inline zebra-network/src/peer_set.rs:149]: 🤝

File: zebra-network/src/peer_set.rs [hunk 5]
```diff
@@ -158,41 +146,33 @@ where
...
-                let stream = TcpStream::connect(addr).await?;
-                pc.ready().await?;
-                let client = pc.call((stream, addr)).await?;
-                Ok::<_, BoxedStdError>(Change::Insert(addr, client))
-            }
-        })
-        .collect::<FuturesUnordered<_>>();
+    use tower::util::CallAllUnordered;
+    let addr_stream = futures::stream::iter(initial_peers.into_iter());
+    let mut handshakes = CallAllUnordered::new(connector, addr_stream);
```
Comment (dconnolly) [inline zebra-network/src/peer_set.rs:164]: neat

**Review by dconnolly — APPROVED** [review]
⛵

**Pull request merged**